File: R/R/recommend.R
#' Field-aware palette recommendation
#'
#' Returns the best-matching palette for the given figure context by
#' cross-referencing the shared \code{data/palettes.yml} and
#' \code{data/field_rules.yml} rule files.
#'
#' @param field Character. Scientific domain: \code{"clinical"},
#'   \code{"omics"}, \code{"singlecell"}, \code{"mechanism"},
#'   or \code{"engineering"}.
#' @param figure_type Character. Chart type: \code{"bar"}, \code{"box"},
#'   \code{"violin"}, \code{"line"}, \code{"scatter"}, \code{"heatmap"},
#'   \code{"volcano"}, or \code{"umap"}.
#' @param variable_type Character. Data encoding: \code{"categorical"},
#'   \code{"sequential"}, or \code{"diverging"}. Default \code{"categorical"}.
#' @param n_groups Integer or NULL. Number of categories or levels.
#' @param colorblind_safe Logical. Restrict to colorblind-safe palettes.
#'   Default \code{FALSE}.
#' @param grayscale_safe Logical. Restrict to grayscale-safe palettes.
#'   Default \code{FALSE}.
#' @return A named list with elements \code{palette_id}, \code{hex},
#'   \code{n_max}, \code{variable_type}, \code{journal_family},
#'   \code{colorblind_safe}, \code{grayscale_safe}, \code{rationale},
#'   \code{warnings}.
#' @export
#' @examples
#' result <- recommend_palette("clinical", "box", n_groups = 4,
#'                             colorblind_safe = TRUE)
#' result$palette_id
#' result$hex
recommend_palette <- function(
    field,
    figure_type,
    variable_type = "categorical",
    n_groups = NULL,
    colorblind_safe = FALSE,
    grayscale_safe = FALSE
) {
  valid_fields <- c("clinical", "omics", "singlecell", "mechanism", "engineering")
  valid_figure_types <- c("bar", "box", "violin", "line", "scatter",
                          "heatmap", "volcano", "umap")
  valid_var_types <- c("categorical", "sequential", "diverging")

  field <- tolower(field)
  figure_type <- tolower(figure_type)
  variable_type <- tolower(variable_type)

  if (!field %in% valid_fields) {
    stop(sprintf("field must be one of: %s", paste(valid_fields, collapse = ", ")))
  }
  if (!figure_type %in% valid_figure_types) {
    stop(sprintf("figure_type must be one of: %s", paste(valid_figure_types, collapse = ", ")))
  }
  if (!variable_type %in% valid_var_types) {
    stop(sprintf("variable_type must be one of: %s", paste(valid_var_types, collapse = ", ")))
  }

  palettes_data <- .load_palettes_yaml()
  field_rules   <- .load_field_rules()

  field_cfg    <- field_rules$fields[[field]]
  preferred_ids <- field_cfg$preferred_palette_ids %||% character(0)

  candidates <- .filter_palette_candidates(
    palettes      = palettes_data$palettes,
    field         = field,
    figure_type   = figure_type,
    variable_type = variable_type,
    n_groups      = n_groups,
    colorblind_safe = colorblind_safe,
    grayscale_safe  = grayscale_safe,
    preferred_ids   = preferred_ids
  )

  if (length(candidates) == 0) {
    stop(sprintf(
      "No palette found for field='%s', figure_type='%s', variable_type='%s'. ",
      field, figure_type, variable_type
    ))
  }

  palette_id <- candidates[[1]]$id
  pal        <- candidates[[1]]$data
  hex        <- unlist(pal$colors)

  if (!is.null(n_groups) && n_groups <= length(hex)) {
    hex <- hex[seq_len(n_groups)]
  }

  list(
    palette_id      = palette_id,
    hex             = hex,
    n_max           = pal$n_max,
    variable_type   = pal$variable_type,
    journal_family  = pal$journal_family,
    colorblind_safe = isTRUE(pal$colorblind_safe),
    grayscale_safe  = isTRUE(pal$grayscale_safe),
    rationale       = trimws(pal$rationale %||% ""),
    warnings        = pal$warnings %||% list()
  )
}


.filter_palette_candidates <- function(palettes, field, figure_type, variable_type,
                                       n_groups, colorblind_safe, grayscale_safe,
                                       preferred_ids) {
  scored <- list()
  for (pid in names(palettes)) {
    pal <- palettes[[pid]]

    if (!is.null(variable_type) && pal$variable_type != variable_type) next
    if (colorblind_safe && !isTRUE(pal$colorblind_safe)) next
    if (grayscale_safe && !isTRUE(pal$grayscale_safe)) next
    if (!is.null(n_groups) && n_groups > (pal$n_max %||% 999L)) next

    score <- 0L
    if (field %in% (pal$field_tags %||% character(0)))       score <- score + 10L
    if (figure_type %in% (pal$figure_tags %||% character(0))) score <- score + 5L
    if (pid %in% preferred_ids)                               score <- score + 8L

    scored <- c(scored, list(list(id = pid, data = pal, score = score)))
  }

  if (length(scored) == 0) return(list())

  scores <- vapply(scored, `[[`, 0L, "score")
  scored[order(-scores)]
}


.load_palettes_yaml <- function() {
  yaml_path <- .find_data_file("palettes.yml")
  yaml::read_yaml(yaml_path)
}

.load_field_rules <- function() {
  yaml_path <- .find_data_file("field_rules.yml")
  yaml::read_yaml(yaml_path)
}

.find_data_file <- function(filename) {
  # When installed as R package
  pkg_path <- system.file("extdata", filename, package = "pubchroma")
  if (nzchar(pkg_path)) return(pkg_path)

  # Development fallback: locate data/ relative to this file
  here <- tryCatch(
    dirname(sys.frame(1)$ofile),
    error = function(e) dirname(normalizePath("R/R/recommend.R", mustWork = FALSE))
  )
  repo_root <- dirname(dirname(here))
  file.path(repo_root, "data", filename)
}

# Null-coalescing helper (avoid purrr dependency)
`%||%` <- function(a, b) if (!is.null(a)) a else b

File: src/pubchroma/recommend.py
"""Field-aware palette recommendation engine.

Usage::

    from pubchroma.recommend import recommend_palette

    result = recommend_palette(
        field="clinical",
        figure_type="box",
        variable_type="categorical",
        n_groups=4,
        colorblind_safe=True,
    )
    print(result["palette_id"])
    print(result["hex"])
    print(result["rationale"])
"""

from __future__ import annotations

from typing import Optional

from ._data_loader import load_field_rules, load_palettes_yaml

# Valid input values
_VALID_FIELDS = {"clinical", "omics", "singlecell", "mechanism", "engineering"}
_VALID_FIGURE_TYPES = {
    "bar", "box", "violin", "line", "scatter", "heatmap", "volcano", "umap",
}
_VALID_VAR_TYPES = {"categorical", "sequential", "diverging"}
_VALID_JOURNAL_FAMILIES = {
    "cns_inspired", "medical_conservative", "engineering_high_contrast", "universal",
}


def recommend_palette(
    field: str,
    figure_type: str,
    variable_type: str = "categorical",
    n_groups: Optional[int] = None,
    journal_family: Optional[str] = None,
    colorblind_safe: bool = False,
    grayscale_safe: bool = False,
) -> dict:
    """Return the best-matching palette for the given figure context.

    Parameters
    ----------
    field : str
        Scientific domain.  One of: ``clinical``, ``omics``, ``singlecell``,
        ``mechanism``, ``engineering``.
    figure_type : str
        Chart type.  One of: ``bar``, ``box``, ``violin``, ``line``,
        ``scatter``, ``heatmap``, ``volcano``, ``umap``.
    variable_type : str, optional
        Data encoding type: ``categorical``, ``sequential``, or ``diverging``.
        Default ``"categorical"``.
    n_groups : int, optional
        Number of categories or levels.  Used to check palette capacity.
    journal_family : str, optional
        Preferred style family.  One of: ``cns_inspired``,
        ``medical_conservative``, ``engineering_high_contrast``, ``universal``.
        If ``None``, the field default is used.
    colorblind_safe : bool, optional
        Restrict to colorblind-safe palettes only.
    grayscale_safe : bool, optional
        Restrict to grayscale-safe palettes only.

    Returns
    -------
    dict
        Keys:
        ``palette_id``, ``hex``, ``n_max``, ``variable_type``,
        ``journal_family``, ``colorblind_safe``, ``grayscale_safe``,
        ``rationale``, ``warnings``,
        ``code_snippet_python``, ``code_snippet_r``.

    Raises
    ------
    ValueError
        If required inputs are invalid or no palette matches the constraints.
    """
    field = field.lower()
    figure_type = figure_type.lower()
    variable_type = variable_type.lower()

    if field not in _VALID_FIELDS:
        raise ValueError(f"field must be one of {sorted(_VALID_FIELDS)}, got '{field}'")
    if figure_type not in _VALID_FIGURE_TYPES:
        raise ValueError(
            f"figure_type must be one of {sorted(_VALID_FIGURE_TYPES)}, got '{figure_type}'"
        )
    if variable_type not in _VALID_VAR_TYPES:
        raise ValueError(
            f"variable_type must be one of {sorted(_VALID_VAR_TYPES)}, got '{variable_type}'"
        )
    if journal_family is not None and journal_family not in _VALID_JOURNAL_FAMILIES:
        raise ValueError(
            f"journal_family must be one of {sorted(_VALID_JOURNAL_FAMILIES)}, "
            f"got '{journal_family}'"
        )

    palettes = load_palettes_yaml()["palettes"]
    field_rules = load_field_rules()["fields"]
    field_cfg = field_rules.get(field, {})

    preferred_ids: list[str] = field_cfg.get("preferred_palette_ids", [])
    preferred_families: list[str] = field_cfg.get("preferred_journal_families", [])

    candidates = _filter_candidates(
        palettes=palettes,
        field=field,
        figure_type=figure_type,
        variable_type=variable_type,
        n_groups=n_groups,
        journal_family=journal_family or (preferred_families[0] if preferred_families else None),
        colorblind_safe=colorblind_safe,
        grayscale_safe=grayscale_safe,
        preferred_ids=preferred_ids,
    )

    if not candidates:
        # Relax constraints progressively
        candidates = _filter_candidates(
            palettes=palettes,
            field=field,
            figure_type=figure_type,
            variable_type=variable_type,
            n_groups=n_groups,
            journal_family=None,
            colorblind_safe=colorblind_safe,
            grayscale_safe=grayscale_safe,
            preferred_ids=preferred_ids,
        )

    if not candidates:
        raise ValueError(
            f"No palette found for field='{field}', figure_type='{figure_type}', "
            f"variable_type='{variable_type}', colorblind_safe={colorblind_safe}, "
            f"grayscale_safe={grayscale_safe}.  "
            "Try relaxing colorblind_safe or grayscale_safe constraints."
        )

    palette_id, pal = candidates[0]
    return _build_result(palette_id, pal, n_groups)


def _filter_candidates(
    palettes: dict,
    field: str,
    figure_type: str,
    variable_type: str,
    n_groups: Optional[int],
    journal_family: Optional[str],
    colorblind_safe: bool,
    grayscale_safe: bool,
    preferred_ids: list[str],
) -> list[tuple[str, dict]]:
    scored: list[tuple[int, str, dict]] = []

    for pid, pal in palettes.items():
        # Hard filters
        if variable_type and pal.get("variable_type") != variable_type:
            continue
        if colorblind_safe and not pal.get("colorblind_safe", False):
            continue
        if grayscale_safe and not pal.get("grayscale_safe", False):
            continue
        if n_groups is not None and n_groups > pal.get("n_max", 999):
            continue

        # Soft scoring (higher = better)
        score = 0
        if field in pal.get("field_tags", []):
            score += 10
        if figure_type in pal.get("figure_tags", []):
            score += 5
        if journal_family and pal.get("journal_family") == journal_family:
            score += 3
        if pid in preferred_ids:
            score += 8

        scored.append((score, pid, pal))

    scored.sort(key=lambda x: -x[0])
    return [(pid, pal) for _, pid, pal in scored]


def _build_result(palette_id: str, pal: dict, n_groups: Optional[int]) -> dict:
    hex_colors = pal["colors"]
    if n_groups is not None:
        hex_colors = hex_colors[:n_groups] if n_groups <= len(hex_colors) else hex_colors

    colors_r = ", ".join(f'"{c}"' for c in hex_colors)

    code_python = (
        f"from pubchroma.recommend import recommend_palette\n"
        f"result = recommend_palette(...)\n"
        f"colors = {list(hex_colors)}\n\n"
        f"# With matplotlib:\n"
        f"import matplotlib.pyplot as plt\n"
        f"from cycler import cycler\n"
        f"plt.rcParams['axes.prop_cycle'] = cycler(color=colors)"
    )

    code_r = (
        f"library(pubchroma)\n"
        f"colors <- c({colors_r})\n\n"
        f"# With ggplot2:\n"
        f"library(ggplot2)\n"
        f"scale_color_manual(values = colors)"
    )

    return {
        "palette_id": palette_id,
        "hex": list(hex_colors),
        "n_max": pal["n_max"],
        "variable_type": pal["variable_type"],
        "journal_family": pal.get("journal_family"),
        "colorblind_safe": pal.get("colorblind_safe", False),
        "grayscale_safe": pal.get("grayscale_safe", False),
        "rationale": pal.get("rationale", "").strip(),
        "warnings": pal.get("warnings", []),
        "code_snippet_python": code_python,
        "code_snippet_r": code_r,
    }
